File: implementations/ruby/string-search/v8-knuth-morris-pratt-d7l8/kmp.rb
def kmp_search(text, pattern)
  return [] if pattern.empty?
  lps = Array.new(pattern.length, 0)
  len = 0
  i = 1
  while i < pattern.length
    if pattern[i] == pattern[len]
      len += 1
      lps[i] = len
      i += 1
    elsif len != 0
      len = lps[len - 1]
    else
      lps[i] = 0
      i += 1
    end
  end
  matches = []
  i = 0
  j = 0
  while i < text.length
    if text[i] == pattern[j]
      i += 1
      j += 1
      if j == pattern.length
        matches << i - j
        j = lps[j - 1]
      end
    elsif j != 0
      j = lps[j - 1]
    else
      i += 1
    end
  end
  matches
end

raise 'ababd' unless kmp_search('ababcabcabababd', 'ababd') == [10]
raise 'aa' unless kmp_search('aaaaa', 'aa') == [0, 1, 2, 3]
raise 'gh' unless kmp_search('abcdef', 'gh') == []
raise 'abc' unless kmp_search('abc', 'abc') == [0]
puts 'ruby kmp ok'

File: implementations/javascript/string-search/v8-knuth-morris-pratt-d7l8/kmp.js
export function kmpSearch(text, pattern) {
  if (pattern.length === 0) return [];
  const lps = Array(pattern.length).fill(0);
  for (let i = 1, len = 0; i < pattern.length;) {
    if (pattern[i] === pattern[len]) {
      lps[i++] = ++len;
    } else if (len !== 0) {
      len = lps[len - 1];
    } else {
      lps[i++] = 0;
    }
  }
  const matches = [];
  for (let i = 0, j = 0; i < text.length;) {
    if (text[i] === pattern[j]) {
      i++;
      j++;
      if (j === pattern.length) {
        matches.push(i - j);
        j = lps[j - 1];
      }
    } else if (j !== 0) {
      j = lps[j - 1];
    } else {
      i++;
    }
  }
  return matches;
}
